sympy/assumptions/cnf.py

```diff
@@ -0,0 +1,373 @@
+from itertools import combinations, product
+
+from sympy import S
+from sympy.logic.boolalg import BooleanFunction, Or, And, Not
+
+
+# Early encoding for literals
+literals_store = dict()
+
+
+class Literal(object):
+    """
+    The smallest element of a CNF object
+    """
+    literals_count = 0
+
+    def __new__(cls, lit, is_not=False):
+        if type(lit).__name__ == 'Not':
+            lit = lit.args[0]
+            is_not = True
+        k = literals_store.get(lit, None)
+        if k is not None:
+            code = k
+        elif lit in (S.false, False):
+            code = 0
+        else:
+            cls.literals_count += 1
+            code = cls.literals_count
+        obj = super(Literal, cls).__new__(cls)
+        obj.code = code
+        obj.lit = lit
+        literals_store[lit] = obj.code
+        obj.is_Not = is_not
+        return obj
+
+    @property
+    def arg(self):
+        sign = -1 if self.is_Not else 1
+        return sign*self.code
+
+    def __invert__(self):
+        is_not = not self.is_Not
+        return Literal(self.lit, is_not=is_not)
+
+    def __str__(self):
+        return '%s( %s, %s)' % (type(self).__name__, self.lit, self.is_Not)
+
+    __repr__ = __str__
+
+    def __eq__(self, other):
+        return self.arg == other.arg
+
+    def __hash__(self):
+        h = hash((type(self).__name__, + self.arg))
+        return h
+
+    def __lt__(self, other):
+        return self.arg < other.arg
+
+
+def literal_sort_key(lit):
+    _type = type(lit).__name__
+    if _type == 'Literal':
+        return lit.arg
+
+    if _type == 'AND':
+        return -2, lit.args
+
+    if _type == 'OR':
+        return -1, lit.args
+
+
+class OR(object):
+    """
+    A low-level implementation for Or
+    """
+    def __init__(self, *args):
+        self._args = args
+
+    @property
+    def args(self):
+        return sorted(self._args, key=literal_sort_key)
+
+    def __invert__(self):
+        return AND(*[~arg for arg in self._args])
+
+    def __hash__(self):
+        return hash((type(self).__name__, ) + tuple(self.args))
+
+    def __eq__(self, other):
+        return self.args == other.args
+
+    def __str__(self):
+        s = '(' + ' | '.join([str(arg) for arg in self.args]) + ')'
+        return s
+
+    __repr__ = __str__
+
+
+class AND(object):
+    """
+    A low-level implementation for And
+    """
+    def __init__(self, *args):
+        self._args = args
+
+    def __invert__(self):
+        return OR(*[~arg for arg in self._args])
+
+    @property
+    def args(self):
+        return sorted(self._args, key=literal_sort_key)
```

Comment:
Does args get used somewhere? This isn't a Basic object, so it shouldn't need to have it. It seems that anything that uses this should just use the unsorted args (perhaps even store them as a set). My worry is that this sorting could slow things down and seems completely unnecessary.

Reply:
Sorted `args` is used for equality and printing. I had this while debugging and I think it's better to have it here. The code doesn't use it, instead `_args` is used.

Reply:
Printing is fine. But I would expect equality to be used in the algorithm somewhere. Can we just store `args` and have it be a set?